searx/network/context.py

```diff
@@ -0,0 +1,321 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
+# pyright: basic
+# pylint: disable=missing-class-docstring
+"""This module impplements NetworkContext which deals with 
+
+* retry strategies: what to do when a HTTP request fails and retries>0
+* record HTTP runtime
+* timeout: In engines, the user query starts at one point in time,
+  the engine timeout is starting point plus a defined value.
+  NetworkContext sends HTTP requests following the engine timeouts.
+
+.. code-block:: python
+        context = NetworkContextRetryFunction(...)  # or another implementation
+
+        def my_engine():
+            http_client = context.get_http_client()
+            ip_ifconfig = http_client.request("GET", "https://ifconfig.me/")
+            print("ip from ifconfig.me ", ip_ifconfig)
+            ip_myip = http_client.request("GET", "https://api.myip.com").json()["ip"]
+            print("ip from api.myip.com", ip_myip)
+            assert ip_ifconfig == ip_myip 
+            # ^^ always true with NetworkContextRetryFunction and NetworkContextRetrySameHTTPClient
+
+        result = context.call(my_engine)
+        print('HTTP runtime:', context.get_total_time())
+
+Note in the code above NetworkContextRetryFunction is instanced directly for the sake of simplicity.
+NetworkContext are instanciated using Network.get_context(...)
+
+Various implementations define what to do when there is an exception in the function `my_engine`:
+
+* `NetworkContextRetryFunction` gets another HTTP client, and try again the whole function.
+* `NetworkContextRetryDifferentHTTPClient` gets another HTTP client, and try again the query.
+* `NetworkContextRetrySameHTTPClient` try again the query with the same HTTP client.
+"""
+import functools
+import ssl
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from timeit import default_timer
+from typing import Callable, Optional, final
+
+try:
+    from typing import ParamSpec, TypeVar
+except ImportError:
+    # to support Python < 3.10
+    from typing_extensions import ParamSpec, TypeVar
+
+import httpx
+
+from searx.network.client import ABCHTTPClient, SoftRetryHTTPException
+
+P = ParamSpec('P')
+R = TypeVar('R')
+HTTPCLIENTFACTORY = Callable[[], ABCHTTPClient]
+
+DEFAULT_TIMEOUT = 120.0
+
+
+## NetworkContext
+
+
+class NetworkContext(ABC):
+    """Abstract implementation: the call must defined in concret classes.
+
+    Life time: one engine request or initialization of an engine
+    """
+
+    __slots__ = ('retries', 'http_client_factory', 'http_client', 'start_time', 'total_time', 'timeout')
+
+    def __init__(
+        self, retries: int, http_client_factory: HTTPCLIENTFACTORY, start_time: Optional[float], timeout: float
```

Comment:
There are two timeout types:
* per request
* per context

The timeout per context should be optional.
It should already work because self.get_remaining_time() already deal with the None value

`timeout: float` --> `timeout: Optional[float]`